flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphRescalingTest.java

```diff
@@ -0,0 +1,235 @@
+		// TODO this limitation should be removed.
+		for (JobVertex jv : jobVertices) {
+			jv.setParallelism(ExecutionConfig.PARALLELISM_AUTO_MAX);
+		}
+
+		ExecutionGraph eg = ExecutionGraphBuilder.buildGraph(
+				null,
+				jobGraph,
+				config,
+				TestingUtils.defaultExecutor(),
+				TestingUtils.defaultExecutor(),
+				new Scheduler(TestingUtils.defaultExecutionContext()),
+				Thread.currentThread().getContextClassLoader(),
+				new StandaloneCheckpointRecoveryFactory(),
+				AkkaUtils.getDefaultTimeout(),
+				new NoRestartStrategy(),
+				new UnregisteredMetricsGroup(),
+				5,
+				TEST_LOGGER);
+
+		for (JobVertex jv : jobVertices) {
+			assertEquals(5, jv.getParallelism());
+		}
+		verifyGeneratedExecutionGraphOfSimpleBitartiteJobGraph(eg, jobVertices);
+
+		// --- verify scaling up works correctly ---
+
+		// TODO rescaling the JobGraph is currently only supported if the
+		// TODO configured parallelism is ExecutionConfig.PARALLELISM_AUTO_MAX.
+		// TODO this limitation should be removed.
+		for (JobVertex jv : jobVertices) {
+			jv.setParallelism(ExecutionConfig.PARALLELISM_AUTO_MAX);
+		}
+
+		eg = ExecutionGraphBuilder.buildGraph(
+				null,
+				jobGraph,
+				config,
+				TestingUtils.defaultExecutor(),
+				TestingUtils.defaultExecutor(),
+				new Scheduler(TestingUtils.defaultExecutionContext()),
+				Thread.currentThread().getContextClassLoader(),
+				new StandaloneCheckpointRecoveryFactory(),
+				AkkaUtils.getDefaultTimeout(),
+				new NoRestartStrategy(),
+				new UnregisteredMetricsGroup(),
+				10,
+				TEST_LOGGER);
+
+		for (JobVertex jv : jobVertices) {
+			assertEquals(10, jv.getParallelism());
+		}
+		verifyGeneratedExecutionGraphOfSimpleBitartiteJobGraph(eg, jobVertices);
+
+		// --- verify down scaling works correctly ---
+
+		// TODO rescaling the JobGraph is currently only supported if the
+		// TODO configured parallelism is ExecutionConfig.PARALLELISM_AUTO_MAX.
+		// TODO this limitation should be removed.
+		for (JobVertex jv : jobVertices) {
+			jv.setParallelism(ExecutionConfig.PARALLELISM_AUTO_MAX);
+		}
+
+		eg = ExecutionGraphBuilder.buildGraph(
+			null,
+			jobGraph,
+			config,
+			TestingUtils.defaultExecutor(),
+			TestingUtils.defaultExecutor(),
+			new Scheduler(TestingUtils.defaultExecutionContext()),
+			Thread.currentThread().getContextClassLoader(),
+			new StandaloneCheckpointRecoveryFactory(),
+			AkkaUtils.getDefaultTimeout(),
+			new NoRestartStrategy(),
+			new UnregisteredMetricsGroup(),
+			2,
+			TEST_LOGGER);
+
+		for (JobVertex jv : jobVertices) {
+			assertEquals(2, jv.getParallelism());
+		}
+		verifyGeneratedExecutionGraphOfSimpleBitartiteJobGraph(eg, jobVertices);
+	}
+
+	/**
+	 * Verifies that building an {@link ExecutionGraph} from a {@link JobGraph} with
+	 * parallelism higher than the maximum parallelism fails.
+	 *
+	 * TODO this test is ignored, since currently the rescale does not properly fail when rescaling to DOP above max.
```

Comment:
What happens if we rescale to something higher?

Reply:
@tillrohrmann currently the rescale doesn't fail, and the created `ExecutionGraph` will have DOP as specified (above max).

Reply:
I mean it makes sense for operators to be scalable beyond max parallelism which don't have keyed state (not sure whether we do this distinction). For operators with keyed state a scaling beyond max parallelism should produce a proper failure message. Maybe we can add this.

Reply:
I like that `maxParallelism` applies to all operators as a constraint.